lib/webdriveragent.js

```diff
@@ -46,123 +48,210 @@ class WebDriverAgent {
+                let msg = `Unable to detect url from WebDriverAgent: ${agentUrl}. ` +
+                          'Is the device on the same network as the server?';
+                log.errorAndThrow(new Error(msg));
+              }
 
-      this.jwproxy = new JWProxy({host: this.url.hostname, port: this.url.port, base: ''});
-      this.jwproxy.sessionId = sessionId;
-      this.proxyReqRes = this.jwproxy.proxyReqRes.bind(this.jwproxy);
+              return true;
+            }
+          }
 
-      return resolve(url);
+          // periodically log, so it does not look like everything died
+          if (lineCount % 100 === 0) {
+            log.debug('Waiting for WebDriverAgent server to finish loading...');
+          }
+
+          return false;
+        });
+
+        log.info('Waiting for WebDriverAgent to start on device');
+        await agentStartedOnDevice;
+        log.info(`WebDriverAgent started at url '${agentUrl}'`);
+
+        this.url = url.parse(agentUrl);
+        this.jwproxy = new JWProxy({server: this.url.hostname, port: this.url.port, base: ''});
+        this.jwproxy.sessionId = sessionId;
+        this.proxyReqRes = this.jwproxy.proxyReqRes.bind(this.jwproxy);
+        return resolve(url);
+      } catch (err) {
+        log.warn(`Starting xcodebuild failed: ${err.message}`);
+        reject(err);
+      }
     });
   }
 
   async checkForDependencies () {
-    if (!await fs.hasAccess(this.bootstrapPath + "/Carthage")) {
+    if (!await fs.hasAccess(`${this.bootstrapPath}/Carthage`)) {
       await exec('/bin/bash', ['Scripts/bootstrap.sh'], {cwd: this.bootstrapPath});
     }
   }
 
-  createXcodeBuildSubProcess () {
+  createXcodeBuildSubProcess (reject, realDevice = false) {
     let args = [
       '-project',
       this.agentPath,
       '-scheme',
       'WebDriverAgentRunner',
+      '-sdk',
+      realDevice ? 'iphoneos' : 'iphonesimulator',
       '-destination',
-      `id=${this.sim.udid}`,
-      'test'
+      `platform=${realDevice ? 'iOS' : 'iOS Simulator'},id=${this.device.udid}`,
+      'test',
+      `CODE_SIGN_IDENTITY=${realDevice ? 'iPhone Developer' : '""'}`,
+      `CODE_SIGNING_REQUIRED=${realDevice ? 'YES' : 'NO'}`
     ];
-    return new SubProcess('xcodebuild', args, {cwd: this.bootstrapPath});
+    let xcodebuild = new SubProcess('xcodebuild', args, {cwd: this.bootstrapPath});
+
+    this.setupXcodeLogging(xcodebuild, reject);
+
+    return xcodebuild;
+  }
+
+  setupXcodeLogging (xcodebuild, reject) {
+    xcodebuild.on('output', (d, e) => {
+      // we want to pull out the log file that is created, and highlight it
+      // for diagnostic purposes
+      if (e.indexOf('Writing diagnostic log for test session to') !== -1) {
+        // pull out the first line that begins with the path separator
+        // which should be the line indicating the log file generated
+        this.logLocation = _.first(_.remove(e.trim().split('\n'), (v) => v.indexOf(path.sep) === 0));
```

Comment:
i've been preferring the model of managing state at as high a level as possible in the application. on that model, we'd return `logLocation` and if it needs to be managed as state, that would happen in the calling method, or that method's calling method, as high as possible. in many cases i've discovered we don't even need to store something on `this`.